File: sdks/go/autonoma/schema.go
package autonoma

import (
	"reflect"
	"strings"
	"time"

	"github.com/google/uuid"
)

// goTypeToSDKType maps a Go reflect.Type to the SDK's coarse type string.
func goTypeToSDKType(t reflect.Type) string {
	// Unwrap pointer types.
	for t.Kind() == reflect.Ptr {
		t = t.Elem()
	}

	// Check well-known types first.
	if t == reflect.TypeOf(time.Time{}) {
		return "timestamp"
	}
	if t == reflect.TypeOf(uuid.UUID{}) {
		return "uuid"
	}

	switch t.Kind() {
	case reflect.Bool:
		return "boolean"
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return "integer"
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return "integer"
	case reflect.Float32, reflect.Float64:
		return "number"
	case reflect.String:
		return "string"
	case reflect.Slice, reflect.Array, reflect.Map, reflect.Struct:
		return "json"
	default:
		return "string"
	}
}

// fieldName returns the JSON field name for a struct field.
// Uses the json struct tag if present, otherwise lowercases the first char.
func fieldName(f reflect.StructField) string {
	tag := f.Tag.Get("json")
	if tag != "" {
		name := strings.Split(tag, ",")[0]
		if name != "" && name != "-" {
			return name
		}
		if name == "-" {
			return ""
		}
	}
	// Fall back to lowercased first char of field name.
	if len(f.Name) == 0 {
		return ""
	}
	return strings.ToLower(f.Name[:1]) + f.Name[1:]
}

// isFieldRequired returns true if the field has no default value capability.
// Pointer types and types with the "omitempty" json tag are considered optional.
func isFieldRequired(f reflect.StructField) bool {
	tag := f.Tag.Get("json")
	if strings.Contains(tag, "omitempty") {
		return false
	}
	if f.Type.Kind() == reflect.Ptr {
		return false
	}
	return true
}

// modelToFields walks a Go struct's fields to produce a list of FieldInfo.
// Every model gets a synthetic "id" field at the head of the list.
func modelToFields(inputStruct reflect.Type) []FieldInfo {
	// Unwrap pointer.
	for inputStruct.Kind() == reflect.Ptr {
		inputStruct = inputStruct.Elem()
	}

	fields := []FieldInfo{
		{
			Name:       "id",
			Type:       "string",
			IsRequired: false,
			IsId:       true,
			HasDefault: true,
		},
	}

	if inputStruct.Kind() != reflect.Struct {
		return fields
	}

	for i := 0; i < inputStruct.NumField(); i++ {
		f := inputStruct.Field(i)
		// Skip unexported fields.
		if !f.IsExported() {
			continue
		}
		name := fieldName(f)
		if name == "" {
			continue
		}

		hasDefault := f.Type.Kind() == reflect.Ptr || strings.Contains(f.Tag.Get("json"), "omitempty")

		fields = append(fields, FieldInfo{
			Name:       name,
			Type:       goTypeToSDKType(f.Type),
			IsRequired: isFieldRequired(f),
			IsId:       false,
			HasDefault: hasDefault,
		})
	}

	return fields
}

// camelToSnake converts "OrgMember" to "org_member" for cosmetic tableName.
func camelToSnake(name string) string {
	var out []byte
	for i, ch := range name {
		if ch >= 'A' && ch <= 'Z' && i > 0 && name[i-1] >= 'a' && name[i-1] <= 'z' {
			out = append(out, '_')
		}
		out = append(out, byte(ch+32*(rune('A')^ch)>>31&0)) // lowercase
	}
	return strings.ToLower(string(out))
}

// BuildSchemaFromFactories builds the SDK's discover-time schema from registered factories.
func BuildSchemaFromFactories(factories FactoryRegistry, scopeField string) (*SchemaInfo, error) {
	var models []ModelInfo

	for entity, factory := range factories {
		if factory.InputStruct == nil {
			return nil, &AutonomaError{
				Message: "Factory \"" + entity + "\" has no InputStruct. Every factory must declare a Go struct type in InputStruct.",
				Code:    "INTERNAL_ERROR",
				Status:  500,
			}
		}
		models = append(models, ModelInfo{
			Name:      entity,
			TableName: camelToSnake(entity),
			Fields:    modelToFields(factory.InputStruct),
		})
	}

	return &SchemaInfo{
		Models:     models,
		Edges:      []FKEdge{},
		Relations:  []SchemaRelation{},
		ScopeField: scopeField,
	}, nil
}

// SchemaToWire serializes a SchemaInfo to the JSON shape the dashboard expects.
func SchemaToWire(schema *SchemaInfo) map[string]any {
	models := make([]map[string]any, len(schema.Models))
	for i, m := range schema.Models {
		fields := make([]map[string]any, len(m.Fields))
		for j, f := range m.Fields {
			fields[j] = map[string]any{
				"name":       f.Name,
				"type":       f.Type,
				"isRequired": f.IsRequired,
				"isId":       f.IsId,
				"hasDefault": f.HasDefault,
			}
		}
		models[i] = map[string]any{
			"name":      m.Name,
			"tableName": m.TableName,
			"fields":    fields,
		}
	}

	edges := make([]map[string]any, len(schema.Edges))
	for i, e := range schema.Edges {
		edges[i] = map[string]any{
			"from":         e.From,
			"to":           e.To,
			"localField":   e.LocalField,
			"foreignField": e.ForeignField,
			"nullable":     e.Nullable,
		}
	}

	relations := make([]map[string]any, len(schema.Relations))
	for i, r := range schema.Relations {
		relations[i] = map[string]any{
			"parentModel": r.ParentModel,
			"childModel":  r.ChildModel,
			"parentField": r.ParentField,
			"childField":  r.ChildField,
		}
	}

	return map[string]any{
		"models":     models,
		"edges":      edges,
		"relations":  relations,
		"scopeField": schema.ScopeField,
	}
}

File: sdks/ruby/lib/autonoma/schema.rb
# frozen_string_literal: true

module Autonoma
  module Schema
    VALID_TYPES = Set.new(%w[string integer number boolean timestamp date uuid json]).freeze

    # Map a Ruby type string to the SDK's coarse type string.
    def self.field_type_from_class(type_str)
      return "string" if type_str.nil?

      normalized = type_str.to_s.downcase
      VALID_TYPES.include?(normalized) ? normalized : "string"
    end

    # Convert CamelCase to snake_case for cosmetic tableName.
    def self.camel_to_snake(name)
      out = []
      name.each_char.with_index do |ch, i|
        if ch =~ /[A-Z]/ && i > 0 && name[i - 1] !~ /[A-Z]/
          out << "_"
        end
        out << ch.downcase
      end
      out.join
    end

    # Build field list from input_fields array.
    # Each entry is a hash with :name, :type, :required keys.
    def self.fields_from_input_fields(input_fields)
      fields = [
        FieldInfo.new(
          name: "id",
          type: "string",
          is_required: false,
          is_id: true,
          has_default: true
        )
      ]

      (input_fields || []).each do |f|
        name = f[:name] || f["name"]
        type = f[:type] || f["type"]
        required = f[:required].nil? ? f["required"] : f[:required]

        fields << FieldInfo.new(
          name: name.to_s,
          type: field_type_from_class(type),
          is_required: !!required,
          is_id: false,
          has_default: !required
        )
      end

      fields
    end

    # Build the SDK's discover-time schema from registered factories.
    def self.build_schema_from_factories(factories, scope_field)
      models = []

      (factories || {}).each do |entity, factory|
        if factory.input_fields.nil? || factory.input_fields.empty?
          raise ArgumentError,
                "Factory \"#{entity}\" has no input_fields. " \
                "Every factory must declare input_fields in define_factory(...)."
        end

        models << ModelInfo.new(
          name: entity,
          table_name: camel_to_snake(entity),
          fields: fields_from_input_fields(factory.input_fields)
        )
      end

      SchemaInfo.new(
        models: models,
        edges: [],
        relations: [],
        scope_field: scope_field
      )
    end

    # Serialise a SchemaInfo to the JSON shape the dashboard expects.
    def self.schema_to_wire(schema)
      {
        "models" => schema.models.map do |m|
          {
            "name" => m.name,
            "tableName" => m.table_name,
            "fields" => m.fields.map do |f|
              {
                "name" => f.name,
                "type" => f.type,
                "isRequired" => f.is_required,
                "isId" => f.is_id,
                "hasDefault" => f.has_default
              }
            end
          }
        end,
        "edges" => schema.edges.map do |e|
          {
            "from" => e.from_model,
            "to" => e.to_model,
            "localField" => e.local_field,
            "foreignField" => e.foreign_field,
            "nullable" => e.nullable
          }
        end,
        "relations" => schema.relations.map do |r|
          {
            "parentModel" => r.parent_model,
            "childModel" => r.child_model,
            "parentField" => r.parent_field,
            "childField" => r.child_field
          }
        end,
        "scopeField" => schema.scope_field
      }
    end

    # Validate input fields hash: strip unknown keys, check required fields.
    def self.validate_input(fields, input_fields)
      known_names = Set.new
      required_names = []

      (input_fields || []).each do |f|
        name = (f[:name] || f["name"]).to_s
        known_names.add(name)
        req = f[:required].nil? ? f["required"] : f[:required]
        required_names << name if req
      end

      # Strip unknown keys
      validated = fields.select { |k, _| known_names.include?(k.to_s) }

      # Check required fields
      missing = required_names.select { |name| !validated.key?(name) && !validated.key?(name.to_sym) }
      unless missing.empty?
        raise Errors.invalid_body("missing required fields: #{missing.join(", ")}")
      end

      validated
    end

    private_class_method :camel_to_snake, :fields_from_input_fields
  end
end
